src/chains/ethereum/ethereum/src/helpers/simulation-handler.ts

```diff
@@ -0,0 +1,479 @@
+import { DefaultStateManager } from "@ethereumjs/vm/dist/state/index";
+import { Address } from "@ganache/ethereum-address";
+import { calculateIntrinsicGas } from "@ganache/ethereum-transaction";
+import { CallError } from "@ganache/ethereum-utils";
+import { BUFFER_EMPTY, Data, hasOwn, keccak, Quantity } from "@ganache/utils";
+import {
+  makeStepEvent,
+  VmAfterTransactionEvent,
+  VmBeforeTransactionEvent,
+  VmConsoleLogEvent,
+  VmStepEvent
+} from "../provider-events";
+import {
+  Address as EthereumJsAddress,
+  BN,
+  KECCAK256_NULL
+} from "ethereumjs-util";
+import { GanacheTrie } from "./trie";
+import Blockchain from "../blockchain";
+import { Block, RuntimeBlock } from "@ganache/ethereum-block";
+import { EVMResult } from "@ethereumjs/vm/dist/evm/evm";
+import {
+  AccessList,
+  AccessLists
+} from "@ganache/ethereum-transaction/src/access-lists";
+import { ERROR, VmError } from "@ethereumjs/vm/dist/exceptions";
+import { warmPrecompiles } from "./precompiles";
+import { maybeGetLogs } from "@ganache/console.log";
+import Emittery from "emittery";
+
+export type SimulationTransaction = {
+  /**
+   * The address the transaction is sent from.
+   */
+  from: Address;
+  /**
+   * The address the transaction is directed to.
+   */
+  to?: Address;
+  /**
+   * Integer of the gas provided for the transaction execution. eth_call consumes zero gas, but this parameter may be needed by some executions.
+   */
+  gas: Quantity;
+  /**
+   * Integer of the gasPrice used for each paid gas
+   */
+  gasPrice: Quantity;
+  /**
+   * Integer of the value sent with this transaction
+   */
+  value?: Quantity;
+  /**
+   * Hash of the method signature and encoded parameters. For details see Ethereum Contract ABI in the Solidity documentation
+   */
+  data?: Data;
+  block: RuntimeBlock;
+  /**
+   * Array of addresses and storage keys.
+   */
+  accessList?: AccessList;
+};
+
+type CallOverride =
+  | Partial<{
+      code: string;
+      nonce: string;
+      balance: string;
+      state: { [slot: string]: string };
+      stateDiff: never;
+    }>
+  | Partial<{
+      code: string;
+      nonce: string;
+      balance: string;
+      state: never;
+      stateDiff: { [slot: string]: string };
+    }>;
+
+export type CallOverrides = {
+  [address: string]: CallOverride;
+};
+
+/**
+ * Stripped down version of the type from EthereumJs
+ */
+interface RunCallOpts {
+  caller?: EthereumJsAddress;
+  data?: Buffer;
+  gasPrice?: BN;
+  gasLimit?: BN;
+  to?: EthereumJsAddress;
+  value?: BN;
+  block?: any;
+}
+
+export default class SimulationHandler extends Emittery<{
+  "ganache:vm:tx:step": VmStepEvent;
+  "ganache:vm:tx:before": VmBeforeTransactionEvent;
+  "ganache:vm:tx:after": VmAfterTransactionEvent;
+  "ganache:vm:tx:console.log": VmConsoleLogEvent;
+}> {
+  #stateTrie: GanacheTrie;
+  #vm: VM;
+  #stateManager: DefaultStateManager;
+  #runCallOpts: RunCallOpts;
+  // List of addresses to be excluded from access list
```

Comment:
Comments like this should probably be in JSDoc format so they can also be read in IDE tooltips/intellisense.

Reply:
Done!